src/Reports/Sonar.php

```diff
@@ -0,0 +1,128 @@
+<?php
+/**
+ * SonarQube report for PHP_CodeSniffer.
+ *
+ * @author  Marek Víger <marek.viger@gmail.com>
+ * @license https://github.com/squizlabs/PHP_CodeSniffer/blob/master/licence.txt BSD Licence
+ */
+
+namespace PHP_CodeSniffer\Reports;
+
+use PHP_CodeSniffer\Files\File;
+
+class Sonar implements Report
+{
+
+
+    /**
+     * Generate a partial report for a single processed file.
+     *
+     * Function should return TRUE if it printed or stored data about the file
+     * and FALSE if it ignored the file. Returning TRUE indicates that the file and
+     * its data should be counted in the grand totals.
+     *
+     * @param array $report      Prepared report data.
+     * @param File  $phpcsFile   The file being reported on.
+     * @param bool  $showSources Show sources?
+     * @param int   $width       Maximum allowed line width.
+     *
+     * @return bool
+     */
+    public function generateFileReport($report, File $phpcsFile, $showSources=false, $width=80)
+    {
+        foreach ($report['messages'] as $line => $lineErrors) {
+            foreach ($lineErrors as $column => $colErrors) {
+                foreach ($colErrors as $error) {
+                    $error['message'] = strtr(
+                        $error['message'],
+                        [
+                            "\n" => '\n',
+                            "\r" => '\r',
+                            "\t" => '\t',
+                        ]
+                    );
+
+                    $issue = [
+                        'engineId'        => 'PHP_CodeSniffer',
+                        'ruleId'          => $error['source'],
+                        'type'            => 'CODE_SMELL',
+                        'severity'        => $this->convertErrorTypeToSonarSeverity($error['type']),
+                        'primaryLocation' => [
+                            'message'   => $error['message'],
+                            'filePath'  => $report['filename'],
+                            'textRange' => [
+                                'startLine'   => $line,
+                                'endLine'     => $line,
+                                'startColumn' => $column,
```

Comment:
The start column should be less than the end column (not the same or more), this causes the sonar-scanner to fail. As far as I know there is no quick and easy way to determine the real start of the line (excluding whitespace) and imho I see no reason why we should. Simply set the `startColumn` to `0` and make sure `endColumn` is always more than `0` (or `startColumn` for that matter). I suggest checking that beforehand and incrementing with 1 if needed.

I tested this with SQ 7.9.1.27448 and with that modification it works fine.

Reply:
thanks @mdeboer I didn't noticed crashes of sonar-scanner.
What about making `endColumn` optional as it's written in documentation?

Reply:
Ah yes, I missed that. I will try that as that'd be a better solution!